src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/Pipeline/ParameterCapturingPipeline.cs

```diff
@@ -0,0 +1,169 @@
+                    continue;
+                }
+
+                using CancellationTokenSource cts = CancellationTokenSource.CreateLinkedTokenSource(stoppingToken);
+                cts.CancelAfter(request.Payload.Duration);
+
+                try
+                {
+                    await request.StopRequest.Task.WaitAsync(cts.Token).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+
+                }
+
+                _probeManager.StopCapturing();
+
+                _callbacks.CapturingStop(request.Payload.RequestId);
+                _ = _allRequests.TryRemove(request.Payload.RequestId, out _);
+            }
+        }
+
+        private bool TryStartCapturing(StartCapturingParametersPayload request)
+        {
+            try
+            {
+                MethodResolver resolver = new();
+                List<MethodInfo> methods = new(request.Methods.Length);
+                List<MethodDescription> methodsFailedToResolve = new();
+
+                for (int i = 0; i < request.Methods.Length; i++)
+                {
+                    MethodDescription methodDescription = request.Methods[i];
+
+                    List<MethodInfo> resolvedMethods = resolver.ResolveMethodDescription(methodDescription);
+                    if (resolvedMethods.Count == 0)
+                    {
+                        methodsFailedToResolve.Add(methodDescription);
+                    }
+
+                    methods.AddRange(resolvedMethods);
+                }
+
+                if (methodsFailedToResolve.Count > 0)
+                {
+                    UnresolvedMethodsExceptions ex = new(methodsFailedToResolve);
+                    throw ex;
+                }
+
+                _probeManager.StartCapturing(methods);
+                _callbacks.CapturingStart(request, methods);
+
+                return true;
+            }
+            catch (UnresolvedMethodsExceptions ex)
+            {
+                _callbacks.FailedToCapture(
+                    request.RequestId,
+                    ParameterCapturingEvents.CapturingFailedReason.UnresolvedMethods,
+                    ex.Message);
+            }
+            catch (Exception ex)
+            {
+                _callbacks.FailedToCapture(
+                    request.RequestId,
+                    ParameterCapturingEvents.CapturingFailedReason.InternalError,
+                    ex.ToString());
+            }
+
+            return false;
+        }
+
+        public bool TryComplete()
+        {
+            return _requestQueue.Writer.TryComplete();
+        }
+
+        public void SubmitRequest(StartCapturingParametersPayload payload)
+        {
+            if (payload.Methods.Length == 0)
+            {
+                throw new ArgumentException(nameof(payload.Methods));
+            }
+
+            CapturingRequest request = new(payload);
+            if (!_allRequests.TryAdd(payload.RequestId, request))
```

Comment:
It seems that in some cases the pipeline returns true/false and talks to the callback (which then talks to the eventsource), while in other cases the pipeline throws and the caller catches and then talks to the eventsource.